File: les3/les3_6.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

def fact(n):
    if n == 1:
        return 1
    return n * fact(n - 1)

print(fact(1))
print(fact(5))
print(fact(15))
print(fact(998))

#栈溢出解决方案，尾递归（和循环效果一样）
def facta(n):
    return facta_iter(n, 1)

def facta_iter(num, product):
    if num == 1:
        return product
    return facta_iter(num - 1, num * product)

print(facta(1))
print(facta(5))
print(facta(15))
print(facta(997))
#print(facta(1000))
#汉诺塔问题

import time

def move(n, a, b, c):
    if n == 1:
        print('move', a, '---->', c)
        return
    move(n - 1, a, c, b)
    print('move', a, '---->', c)
    move(n - 1, b, a, c)
t_start = time.clock()
move(20, 'A', 'B', 'C')
t_end = time.clock()
print(t_end - t_start)

File: les3/les3_6.c
/*
 *
 * 这段代码是用C写一个简单的汉诺塔解决问题，并且记录时间
 *
 * */
#include<stdio.h>
#include<time.h>

void move(int n, char a, char b, char c)
{
	if(n == 1){
		printf("move %c ----> %c\n", a, c);
		return;
	}
	move(n - 1, a, c, b);
	printf("move %c ----> %c\n", a, c);
	move(n - 1, b, a, c);
}

int main()
{
	time_t t_start, t_end;
	int n = 50;
	char a, b, c;
	t_start = clock();
	move(n, 'A', 'B', 'C');
	t_end = clock();
	printf("Time: %.0f s\n", difftime(t_end, t_start));
	return 0;
}
